frostsnap_coordinator/src/usb_serial_manager.rs

```diff
@@ -183,178 +152,212 @@ impl Ports {
+                            id = announce.from.to_string(),
+                            "Announced!"
+                        );
+                    }
+                    DeviceSendMessage::Debug { message, device } => {
+                        event!(
+                            Level::DEBUG,
+                            port = serial_number,
+                            from = device.to_string(),
+                            name = self
+                                .device_labels
+                                .get(&device)
+                                .cloned()
+                                .unwrap_or("<unknown>".into()),
+                            message
+                        );
+                    }
+                    DeviceSendMessage::Core(msg) => device_to_coord_msg.push(msg),
+                },
             }
         }
 
-        for (device_id, serial_number) in self.device_ports.clone() {
+        let mut outbox = core::mem::take(&mut self.port_outbox);
+
+        for (device_id, _) in &self.device_ports {
             if self.registered_devices.contains(&device_id) {
                 continue;
             }
 
             if let Some(device_label) = self.device_labels.get(&device_id) {
-                let wrote_ack = {
-                    let device_port = self.ready.get_mut(&serial_number).expect("must exist");
+                outbox.push(DeviceReceiveMessage {
+                    message_body: DeviceReceiveBody::AnnounceAck {
+                        device_label: device_label.to_string(),
+                    },
+                    target_destinations: BTreeSet::from([*device_id]),
+                });
 
-                    device_port.send_message(DeviceReceiveSerial::Message(DeviceReceiveMessage {
-                        message_body: DeviceReceiveBody::AnnounceAck {
-                            device_label: device_label.to_string(),
-                        },
-                        target_destinations: BTreeSet::from([device_id]),
-                    }))
-                };
+                event!(
+                    Level::INFO,
+                    device_id = device_id.to_string(),
+                    "Registered device"
+                );
+                self.registered_devices.insert(*device_id);
+                newly_registered.insert(*device_id);
+            }
+        }
 
-                match wrote_ack {
-                    Ok(_) => {
+        outbox.retain_mut(|send| {
+            let mut ports_to_send_on = HashSet::new();
+            let mut wire_message = send.clone();
+            wire_message.target_destinations.clear();
+
+            send.target_destinations.retain(|destination| {
+                match self.device_ports.get(destination) {
+                    Some(port) => {
+                        ports_to_send_on.insert(port.clone());
+                        wire_message.target_destinations.insert(*destination);
```

Comment:
Does this `wire_message` accumulate new target destinations, what if the target destinations are on different ports?

Would the `wire_message` have both target destinations and get sent along both serial ports?

Reply:
> Does this `wire_message` accumulate new target destinations, what if the target destinations are on different ports?
> 
> Would the `wire_message` have both target destinations and get sent along both serial ports?

It sends the same wire message on all the relevant ports (but i didn't test this because I only have one USB cable!).

Reply:
Tested 👍